File: Rev.0/firmware/lib/src/encoding/BER/decode.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <encoding/BER/BER.h>

extern void fields_free(fields);
extern fields fields_add(fields, field *);

fields unpack(const uint8_t *message, int N);

// clang-format off
field *unpack_integer (const uint8_t *, int, int *);
field *unpack_octets  (const uint8_t *, int, int *);
field *unpack_null    (const uint8_t *, int, int *);
field *unpack_OID     (const uint8_t *, int, int *);
field *unpack_sequence(const uint8_t *, int, int *);
field *unpack_PDU     (const uint8_t *, int, int *);
// clang-format on

uint32_t unpack_length(const uint8_t *message, int N, int *ix);

struct packet BER_decode(const uint8_t *message, int N) {
    packet p = {
        .tag = PACKET_GET,
        .get = {
            .version = 0,
            .request_id = 1,
        },
    };

    // ... unpack message
    fields fields = unpack(message, N);

    if (fields != NULL) {
        fields_free(fields);
    }

    return p;
}

fields unpack(const uint8_t *bytes, int N) {
    fields list;
    field *f;
    int length = 0;
    int ix = 0;

    if ((list = calloc(1, sizeof(fieldp))) == NULL) {
        return NULL;
    } else {
        list[0] = NULL;
    }

    while (ix < N && ++length < 16) {
        uint8_t tag = bytes[ix];
        ix += 1;

        switch (tag) {
        case FIELD_INTEGER:
            if ((f = unpack_integer(bytes, N, &ix)) != NULL) {
                list = fields_add(list, f);
            }
            break;

        case FIELD_OCTET_STRING:
            if ((f = unpack_octets(bytes, N, &ix)) != NULL) {
                list = fields_add(list, f);
            }
            break;

        case FIELD_NULL:
            if ((f = unpack_null(bytes, N, &ix)) != NULL) {
                list = fields_add(list, f);
            }
            break;

        case FIELD_OID:
            if ((f = unpack_OID(bytes, N, &ix)) != NULL) {
                list = fields_add(list, f);
            }
            break;

        case FIELD_SEQUENCE:
            if ((f = unpack_sequence(bytes, N, &ix)) != NULL) {
                list = fields_add(list, f);
            }
            break;

        case FIELD_PDU:
            if ((f = unpack_PDU(bytes, N, &ix)) != NULL) {
                list = fields_add(list, f);
            }
            break;

        default:
            printf("::unknown:%2d  N:%d  ix:%d\n", tag, N, ix);
            ix = N;
        }
    }

    fflush(stdout);

    return list;
}

field *unpack_integer(const uint8_t *message, int N, int *ix) {
    uint32_t length = unpack_length(message, N, ix);
    int64_t value = 0;

    if (length <= 8) {
        for (int i = 0; i < length; i++) {
            value <<= 8;
            value |= message[*ix + i] & 0x00ff;
        }
    }

    printf("::integer     N:%d  ix:%-3d length:%lu  value:%lld\n", N, *ix, length, value);

    // ... compose field
    field *f = (field *)calloc(1, sizeof(field));
    f->tag = FIELD_INTEGER;
    f->integer.value = value;

    *ix += length;

    return f;
}

field *unpack_octets(const uint8_t *message, int N, int *ix) {
    uint32_t length = unpack_length(message, N, ix);
    uint8_t *octets = (uint8_t *)calloc(length, sizeof(uint8_t));

    memmove(octets, &message[*ix], length);

    printf("::octets      N:%d  ix:%-3d length:%lu  octets:%s\n", N, *ix, length, strndup(octets, length));

    // ... compose field
    field *f = (field *)calloc(1, sizeof(field));
    f->tag = FIELD_OCTET_STRING;
    f->string.length = length;
    f->string.octets = octets;

    *ix += length;

    return f;
}

field *unpack_null(const uint8_t *message, int N, int *ix) {
    uint32_t length = unpack_length(message, N, ix);

    printf("::null        N:%d  ix:%-3d length:%lu\n", N, *ix, length);

    // ... compose field
    field *f = (field *)calloc(1, sizeof(field));
    f->tag = FIELD_NULL;

    *ix += length;

    return f;
}

field *unpack_OID(const uint8_t *message, int N, int *ix) {
    uint32_t length = unpack_length(message, N, ix);
    char *OID = (char *)calloc(256, sizeof(char));
    char *p = OID;
    char *end = p + 256;

    p += snprintf(p, end - p, "0");

    if (length > 0 && p < end) {
        p += snprintf(p, end - p, ".%d", message[*ix] / 40);
    }

    if (length > 0 && p < end) {
        p += snprintf(p, end - p, ".%d", message[*ix] % 40);
    }

    int i = 1;
    while (i < length && p < end) {
        uint64_t suboid = 0;

        while (i < length && p < end) {
            uint8_t b = message[*ix + i++];

            suboid <<= 7;
            suboid += (uint64_t)(b & 0x7f);

            if ((b & 0x80) == 0x00) {
                p += snprintf(p, end - p, ".%llu", suboid);
                break;
            }
        }
    }

    printf("::OID         N:%d  ix:%-3d length:%lu  OID:%s\n", N, *ix, length, OID);

    // ... compose field
    field *f = (field *)calloc(1, sizeof(field));
    f->tag = FIELD_OID;
    f->OID.OID = OID;

    *ix += length;

    return f;
}

field *unpack_sequence(const uint8_t *message, int N, int *ix) {
    uint32_t length = unpack_length(message, N, ix);

    printf("::sequence    N:%d  ix:%-3d length:%lu\n", N, *ix, length);

    fields list = unpack(&message[*ix], length);

    printf("::sequence/fields %p\n", list);
    fields_free(list);

    // ... 'k, done'
    field *f = (field *)calloc(1, sizeof(field));
    f->tag = FIELD_SEQUENCE;
    f->sequence.fields = NULL;

    *ix += length;

    return f;
}

field *unpack_PDU(const uint8_t *message, int N, int *ix) {
    uint32_t length = unpack_length(message, N, ix);

    printf("::PDU         N:%d  ix:%-3d length:%lu\n", N, *ix, length);

    fields list = unpack(&message[*ix], length);
    fields_free(list);

    field *f = (field *)calloc(1, sizeof(field));
    f->tag = FIELD_PDU;
    f->PDU.fields = NULL;

    *ix += length;

    return f;
}

uint32_t unpack_length(const uint8_t *message, int N, int *ix) {
    uint32_t length = 0;

    if (*ix < N) {
        uint8_t b = message[*ix];
        *ix += 1;

        if ((b & 0x80) == 0x00) {
            length = (uint32_t)(b & 0x7f);
        } else {
            int len = b & 0x7f;
            for (int i = 0; i < len && *ix < N; i++) {
                length <<= 8;
                length += (uint32_t)(message[*ix]);
                *ix += 1;
            }
        }
    }

    return length;
}

File: Rev.0/zero2w/driver/rpcd/encoding/ASN.1/decode.go
package BER

import (
	"fmt"
)

func Decode(bytes []byte) (any, error) {
	if v, _, err := unpack(bytes); err != nil {
		return nil, err
	} else if v == nil {
		return nil, nil
	} else if seq, ok := v.(sequence); !ok {
		return nil, fmt.Errorf("invalid packet")
	} else if len(seq) < 3 {
		return nil, fmt.Errorf("invalid packet")
	} else if version, ok := seq[0].(int64); !ok {
		return nil, fmt.Errorf("invalid packet version")
	} else if community, ok := seq[1].(string); !ok {
		return nil, fmt.Errorf("invalid packet community")
	} else if PDU, ok := seq[2].(pdu); !ok || len(PDU) < 4 {
		return nil, fmt.Errorf("invalid packet PDU")
	} else if requestID, ok := PDU[0].(int64); !ok {
		return nil, fmt.Errorf("invalid packet request ID")
	} else if errCode, ok := PDU[1].(int64); !ok {
		return nil, fmt.Errorf("invalid packet error field")
	} else if errIndex, ok := PDU[2].(int64); !ok {
		return nil, fmt.Errorf("invalid packet error index")
	} else if varbindlist, ok := PDU[3].(sequence); !ok || len(varbindlist) < 1 {
		return nil, fmt.Errorf("invalid packet variable bind list")
	} else if vars, ok := varbindlist[0].(sequence); !ok || len(vars) < 2 {
		return nil, fmt.Errorf("invalid packet variable list")
	} else if oid, ok := vars[0].(OID); !ok {
		return nil, fmt.Errorf("invalid packet variable OID")
	} else if _, ok := vars[1].(null); !ok {
		return nil, fmt.Errorf("invalid packet variable value")
	} else {
		return GetRequest{
			Version:    uint8(version),
			Community:  community,
			RequestID:  uint32(requestID),
			Error:      errCode,
			ErrorIndex: errIndex,
			OID:        oid,
		}, nil
	}
}

func unpack(bytes []byte) (any, []byte, error) {
	if len(bytes) == 0 {
		return nil, bytes, nil
	}

	for len(bytes) > 0 {
		tag := bytes[0]
		bytes = bytes[1:]

		switch tag {
		case tagInteger:
			return unpack_integer(bytes)

		case tagOctetString:
			return unpack_octets(bytes)

		case tagNull:
			return unpack_null(bytes)

		case tagOID:
			return unpack_OID(bytes)

		case tagPDU:
			return unpack_PDU(bytes)

		case tagSequence:
			return unpack_sequence(bytes)

		default:
			return nil, bytes, fmt.Errorf("unknown tag (%v)", tag)
		}
	}

	return nil, bytes, fmt.Errorf("invalid packet")
}

func unpack_length(bytes []byte) (uint32, []byte, error) {
	if len(bytes) > 0 {
		b := bytes[0]
		bytes = bytes[1:]

		if (b & 0x80) == 0x00 {
			return uint32(b & 0x7f), bytes, nil
		} else {
			length := uint32(0)
			N := uint32(b & 0x7f)
			bytes = bytes[1:]
			for ; N > 0 && len(bytes) > 0; N-- {
				length <<= 8
				length += (uint32)(bytes[0])
				bytes = bytes[1:]
			}

			if N == 0 {
				return length, bytes, nil
			}
		}
	}

	return 0, bytes, fmt.Errorf("invalid field length")
}

func unpack_integer(bytes []byte) (int64, []byte, error) {
	if N, bytes, err := unpack_length(bytes); err != nil {
		return 0, nil, err
	} else if N > uint32(len(bytes)) || N > 8 {
		return 0, nil, fmt.Errorf("invalid integer field")
	} else {
		value := int64(0)
		chunk := bytes[:N]

		if len(chunk) > 0 {
			u8 := uint8(chunk[0])
			chunk = chunk[1:]

			if (u8 & 0x80) == 0x80 {
				value = -1
			} else {
				value = 0
			}

			value <<= 8
			value |= int64(u8) & 0x00ff
		}

		for len(chunk) > 0 {
			u8 := uint8(chunk[0])
			chunk = chunk[1:]

			value <<= 8
			value |= int64(u8) & 0x00ff
		}

		return value, bytes[N:], nil
	}
}

func unpack_octets(bytes []byte) (string, []byte, error) {
	if N, bytes, err := unpack_length(bytes); err != nil {
		return "", nil, err
	} else if N > uint32(len(bytes)) {
		return "", nil, fmt.Errorf("invalid octets field")
	} else {
		return string(bytes[0:N]), bytes[N:], nil
	}
}

func unpack_null(bytes []byte) (null, []byte, error) {
	if N, bytes, err := unpack_length(bytes); err != nil {
		return null{}, nil, err
	} else if N > uint32(len(bytes)) {
		return null{}, nil, fmt.Errorf("invalid null field")
	} else {
		return null{}, bytes[N:], nil
	}
}

func unpack_OID(bytes []byte) (OID, []byte, error) {
	if N, bytes, err := unpack_length(bytes); err != nil {
		return nil, nil, err
	} else if N > uint32(len(bytes)) {
		return nil, nil, err
	} else {
		oid := []uint32{0}
		chunk := bytes[:N]

		if len(chunk) > 0 {
			oid = append(oid, uint32(chunk[0]/40))
			oid = append(oid, uint32(chunk[0]%40))
			chunk = chunk[1:]
		}

		for len(chunk) > 0 {
			v := uint32(0)

			for len(chunk) > 0 {
				b := chunk[0]
				chunk = chunk[1:]

				v <<= 7
				v += (uint32)(b & 0x7f)

				if (b & 0x80) == 0x00 {
					oid = append(oid, v)
					break
				}
			}
		}

		return OID(oid), bytes[N:], nil
	}
}

func unpack_PDU(bytes []byte) (pdu, []byte, error) {
	if N, bytes, err := unpack_length(bytes); err != nil {
		return nil, bytes, err
	} else if N > uint32(len(bytes)) {
		return nil, nil, fmt.Errorf("invalid PDU")
	} else {
		PDU := pdu{}
		chunk := bytes[:N]

		for len(chunk) > 0 {
			if field, remaining, err := unpack(chunk); err != nil {
				return nil, bytes[N:], err
			} else {
				PDU = append(PDU, field)
				chunk = remaining
			}
		}

		return PDU, bytes[N:], nil
	}
}

func unpack_sequence(bytes []byte) (sequence, []byte, error) {
	if N, bytes, err := unpack_length(bytes); err != nil {
		return nil, nil, err
	} else if N > uint32(len(bytes)) {
		return nil, nil, fmt.Errorf("invalid sequence field")
	} else {
		seq := sequence{}
		chunk := bytes[:N]

		for len(chunk) > 0 {
			if field, remaining, err := unpack(chunk); err != nil {
				return nil, nil, err
			} else {
				seq = append(seq, field)
				chunk = remaining
			}
		}

		return seq, bytes[N:], nil
	}
}
